File: pkg/models/provider.go
package models

import (
	"encoding/json"
	"fmt"
	"net/url"
	"strings"
	"time"
)

// ProviderType LLM 提供商类型。
type ProviderType string

// noinspection GoUnusedConst

const (
	ProviderKimi        ProviderType = "kimi"
	ProviderOpenAI      ProviderType = "openai"
	ProviderQwen        ProviderType = "qwen"
	ProviderSiliconFlow ProviderType = "siliconflow"
	ProviderOllama      ProviderType = "ollama"
	ProviderLocal       ProviderType = "local"
	ProviderMicrosoft   ProviderType = "microsoft"
	ProviderGitHub      ProviderType = "github"
	ProviderClaude      ProviderType = "claude"
	ProviderGrok        ProviderType = "grok"
	ProviderDoubao      ProviderType = "doubao"
	ProviderGLM         ProviderType = "zhipu"
	ProviderDeepSeek    ProviderType = "deepseek"
	ProviderMiniMax     ProviderType = "minimax"
	ProviderXiaomi      ProviderType = "xiaomi"
	ProviderHunyuan     ProviderType = "hunyuan"
	ProviderVertex      ProviderType = "vertex"
)

// AuthMethod Provider 认证方式。
type AuthMethod string

// noinspection GoUnusedConst

const (
	AuthMethodAPIToken       AuthMethod = "api_key"
	AuthMethodCLIToken       AuthMethod = "cli_token"
	AuthMethodOAuthDevice    AuthMethod = "oauth_device"
	AuthMethodServiceAccount AuthMethod = "service_account"
)

// AuthParams 各认证方式特有配置参数。
type AuthParams struct {
	APIKey string `json:"api_key,omitempty"`

	CLICredentialPath string `json:"cli_credential_path,omitempty"`

	OAuthClientID string `json:"oauth_client_id,omitempty"`
	OAuthAuthURL  string `json:"oauth_auth_url,omitempty"`
	OAuthTokenURL string `json:"oauth_token_url,omitempty"`
	// OAuthRefreshToken 不序列化到数据库，运行时内存持有
	OAuthRefreshToken string `json:"-"`
	// OAuthAccessToken 不序列化到数据库，运行时内存持有
	OAuthAccessToken string `json:"-"`
	OAuthExpiresAt   int64  `json:"oauth_expires_at,omitempty"`

	GCPProjectID string `json:"gcp_project_id,omitempty"`
	GCPRegion    string `json:"gcp_region,omitempty"`
	SAJSON       string `json:"sa_json,omitempty"`
}

// ProviderModel 表示某个服务商下的单个模型配置。
type ProviderModel struct {
	ID               string `json:"id"`
	Name             string `json:"name"`
	Enabled          bool   `json:"enabled"`
	MaxContextLength int    `json:"maxContextLength,omitempty"`
}

// ProviderConfig LLM Provider 完整配置。
// json tag 使用 camelCase 以匹配前端类型定义和 Wails 绑定生成。
type ProviderConfig struct {
	ID          string          `json:"id"`
	Name        string          `json:"name"`
	Type        ProviderType    `json:"type"`
	APIHost     string          `json:"apiHost"`
	APIKey      string          `json:"apiKey,omitempty"`
	ModelID     string          `json:"modelId"`
	Models      []ProviderModel `json:"models"`
	Temperature float64         `json:"temperature"`
	TimeoutMs   int             `json:"timeoutMs"`
	MaxRetries  int             `json:"maxRetries"`
	MaxTokens   int             `json:"maxTokens"`
	GroupName   string          `json:"group"`
	Enabled     bool            `json:"enabled"`
	SortOrder   int             `json:"sortOrder"`
	CreatedAt   int64           `json:"createdAt"`
	UpdatedAt   int64           `json:"updatedAt"`

	AuthMethod AuthMethod `json:"auth_method"`
	AuthParams AuthParams `json:"auth_params"`
}

// Validate 根据 AuthMethod 校验必填字段。
func (p *ProviderConfig) Validate() error {
	if err := p.validateBaseFields(); err != nil {
		return fmt.Errorf("failed to validate base fields: %w", err)
	}
	return p.validateAuthParams()
}

// validateBaseFields 校验 Provider 通用基础字段。
func (p *ProviderConfig) validateBaseFields() error {
	if p.ID == "" {
		return fmt.Errorf("provider id is required")
	}
	if p.Name == "" {
		return fmt.Errorf("provider name is required")
	}
	if p.APIHost == "" {
		return fmt.Errorf("provider api_host is required")
	}
	u, err := url.Parse(p.APIHost)
	if err != nil || (u.Scheme != "http" && u.Scheme != "https") {
		return fmt.Errorf("provider api_host must be a valid http(s) URL")
	}
	// 向后兼容：有 Models 列表时可不填 ModelID（自动取第一个启用模型）
	if p.ModelID == "" && len(p.Models) == 0 {
		return fmt.Errorf("provider model_id is required")
	}
	if p.Temperature < 0 || p.Temperature > 2 {
		return fmt.Errorf("provider temperature must be in range [0, 2]")
	}
	return nil
}

// validateAuthParams 根据 AuthMethod 校验认证参数。
func (p *ProviderConfig) validateAuthParams() error {
	method := p.AuthMethod
	if method == "" {
		method = AuthMethodAPIToken
	}

	switch method {
	case AuthMethodAPIToken:
		return validateAPIToken(p.APIKey)
	case AuthMethodCLIToken:
		return validateCLIToken(p.AuthParams.CLICredentialPath)
	case AuthMethodOAuthDevice:
		return validateOAuthDevice(p.AuthParams.OAuthClientID, p.AuthParams.OAuthTokenURL)
	case AuthMethodServiceAccount:
		return validateServiceAccount(p.AuthParams.GCPProjectID, p.AuthParams.SAJSON)
	default:
		return fmt.Errorf("unknown auth method: %s", method)
	}
}

func validateAPIToken(apiKey string) error {
	if apiKey == "" {
		return fmt.Errorf("provider api_key is required for auth method api_key")
	}
	return nil
}

func validateCLIToken(path string) error {
	if path == "" {
		return fmt.Errorf("provider cli_credential_path is required for auth method cli_token")
	}
	return nil
}

func validateOAuthDevice(clientID, tokenURL string) error {
	if clientID == "" {
		return fmt.Errorf("provider oauth_client_id is required for auth method oauth_device")
	}
	if tokenURL == "" {
		return fmt.Errorf("provider oauth_token_url is required for auth method oauth_device")
	}
	return nil
}

func validateServiceAccount(projectID, saJSON string) error {
	if projectID == "" {
		return fmt.Errorf("provider gcp_project_id is required for auth method service_account")
	}
	if saJSON == "" {
		return fmt.Errorf("provider sa_json is required for auth method service_account")
	}
	return nil
}

// ResolveAuthToken 根据 AuthMethod 获取 HTTP 认证 Token。
func (p *ProviderConfig) ResolveAuthToken() (string, error) {
	method := p.AuthMethod
	if method == "" {
		method = AuthMethodAPIToken
	}

	switch method {
	case AuthMethodAPIToken:
		// 允许空 API Key（如本地 Ollama 无需认证），由调用方决定是否发送 Authorization
		return p.APIKey, nil
	case AuthMethodCLIToken:
		token, hint, err := ReadCLITokenFromFile(p.AuthParams.CLICredentialPath)
		if err != nil {
			return "", fmt.Errorf("failed to resolve cli token: %w", err)
		}
		if hint == "" {
			// 文件直接包含可用的 access_token
			return token, nil
		}
		// hint == "refresh_token"，检查内存缓存的 access_token
		if p.AuthParams.OAuthAccessToken != "" && p.AuthParams.OAuthExpiresAt > time.Now().Unix() {
			return p.AuthParams.OAuthAccessToken, nil
		}
		return "", fmt.Errorf("access_token expired, refresh required")
	case AuthMethodOAuthDevice:
		// 优先检查内存缓存的 access_token
		if p.AuthParams.OAuthAccessToken != "" && p.AuthParams.OAuthExpiresAt > time.Now().Unix() {
			return p.AuthParams.OAuthAccessToken, nil
		}
		return "", fmt.Errorf("access_token expired, refresh required")
	case AuthMethodServiceAccount:
		return "", fmt.Errorf("service_account auth not yet implemented")
	default:
		return "", fmt.Errorf("unknown auth method: %s", method)
	}
}

// MarshalAuthParams 序列化为 JSON 字符串供数据库存储。
func (p *ProviderConfig) MarshalAuthParams() (string, error) {
	b, err := json.Marshal(p.AuthParams)
	if err != nil {
		return "", fmt.Errorf("failed to marshal auth params: %w", err)
	}
	return string(b), nil
}

// UnmarshalAuthParams 从 JSON 字符串反序列化。
func (p *ProviderConfig) UnmarshalAuthParams(data string) error {
	if data == "" || data == "{}" {
		p.AuthParams = AuthParams{}
		return nil
	}
	if err := json.Unmarshal([]byte(data), &p.AuthParams); err != nil {
		return fmt.Errorf("failed to unmarshal auth params: %w", err)
	}
	return nil
}

// InferProviderType 根据 API Host 推断 ProviderType。
// 用于后端在仅有 api_host 时还原 provider 类型，辅助本地/云端判据。
// 无法识别时返回空字符串。
func InferProviderType(apiHost string) ProviderType {
	u, err := url.Parse(apiHost)
	if err != nil {
		return ""
	}

	host := u.Hostname()
	port := u.Port()

	// 本地回环端点：按常用端口推断 Ollama / 通用本地，其余统一视为 local。
	if host == "localhost" || host == "127.0.0.1" || host == "::1" {
		switch port {
		case "11434":
			return ProviderOllama
		case "8080":
			return ProviderLocal
		default:
			return ProviderLocal
		}
	}

	// 已知云端厂商默认域名。
	knownCloudHosts := map[string]ProviderType{
		"api.moonshot.cn":                   ProviderKimi,
		"api.openai.com":                    ProviderOpenAI,
		"dashscope.aliyuncs.com":            ProviderQwen,
		"api.siliconflow.cn":                ProviderSiliconFlow,
		"models.inference.ai.azure.com":     ProviderMicrosoft,
		"api.anthropic.com":                 ProviderClaude,
		"api.x.ai":                          ProviderGrok,
		"ark.cn-beijing.volces.com":         ProviderDoubao,
		"open.bigmodel.cn":                  ProviderGLM,
		"api.deepseek.com":                  ProviderDeepSeek,
		"api.minimax.chat":                  ProviderMiniMax,
		"api.mi.ai":                         ProviderXiaomi,
		"hunyuan.tencentcloudapi.com":       ProviderHunyuan,
		"generativelanguage.googleapis.com": ProviderKimi, // 未定义 gemini，保守归为 kimi 云端
	}

	for h, pt := range knownCloudHosts {
		if strings.Contains(host, h) {
			return pt
		}
	}

	return ""
}

// ModelConfig 单个模型配置信息。
type ModelConfig struct {
	Provider    ProviderType `json:"provider"`
	ModelName   string       `json:"model_name"`
	APIEndpoint string       `json:"api_endpoint"`
	MaxTokens   int          `json:"max_tokens"`
	Temperature float64      `json:"temperature"`
}

File: web/src/types/provider.ts
/**
 * Provider 模板类型定义。
 * 对应 assets/provider-templates.json 中的静态模板数据。
 */
export interface ProviderTemplate {
  id: string
  name: string
  apiHost: string
  defaultModel: string
  models: string[]
  description: string
  docsUrl: string
  type: 'cloud' | 'local'
  authMethods: AuthMethod[]
}

/**
 * 认证方式枚举。
 * 对应后端 models.AuthMethod。
 */
export type AuthMethod = 'api_key' | 'cli_token' | 'oauth_device' | 'service_account'

/**
 * 各认证方式特有的配置参数。
 */
export interface AuthParams {
  /** API Key 方式（api_key） */
  apiKey?: string
  /** CLI Token 方式（cli_token） */
  cliCredentialPath?: string
  /** OAuth Device Flow 方式（oauth_device） */
  oauthClientId?: string
  oauthAuthUrl?: string
  oauthTokenUrl?: string
  oauthRefreshToken?: string
  oauthAccessToken?: string
  oauthExpiresAt?: number
  /** Service Account 方式（service_account） */
  gcpProjectId?: string
  gcpRegion?: string
  saJson?: string
}

/**
 * CLI 检测结果，对应后端 auth.CLIDetectResult。
 */
export interface CLIDetectResult {
  /** 提供商类型：kimi | gemini */
  providerType: string
  /** CLI 凭证文件是否存在 */
  detected: boolean
  /** 检测到的凭证文件路径 */
  credentialPath: string
  /** 文件内容是否包含有效 token 信息 */
  loggedIn: boolean
  /** 检测过程中的错误提示 */
  error?: string
}

/**
 * 认证降级事件，对应后端 runtime.EventsEmit("auth:degraded", ...)。
 */
export interface AuthDegradedEvent {
  /** 触发降级的 Provider ID */
  providerID: string
  /** 降级原因 */
  reason: string
}

/**
 * OAuth Device Flow 启动响应，对应后端 DeviceFlowStartResponse。
 */
export interface DeviceFlowStartResponse {
  /** 用户码，供前端展示 */
  userCode: string
  /** 验证 URL，供用户浏览器访问 */
  verificationURI: string
  /** 设备码，后端轮询用 */
  deviceCode: string
  /** 设备码有效期（秒） */
  expiresIn: number
  /** 建议轮询间隔（秒） */
  interval: number
  /** 本地回调服务器地址（可选，用于同设备授权后的自动通知） */
  redirectURI?: string
}

/**
 * OAuth Device Flow 状态响应，对应后端 DeviceFlowStatusResponse。
 */
export interface DeviceFlowStatusResponse {
  /** 设备码 */
  deviceCode: string
  /** 厂商类型 */
  providerType: string
  /** 当前状态：pending / slow_down / success / error / cancelled */
  status: string
  /** 错误信息 */
  error?: string
  /** 授权成功后生成的 Provider ID */
  providerID?: string
  /** 授权成功后生成的 Provider 名称 */
  providerName?: string
}

/**
 * OAuth Device Flow 支持的厂商信息，对应后端 OAuthDeviceFlowProviderInfo。
 */
export interface OAuthDeviceFlowProviderInfo {
  /** 厂商类型 */
  providerType: string
  /** 显示名称 */
  name: string
  /** 是否可用（环境变量已配置） */
  available: boolean
  /** 环境变量是否已配置 */
  configured: boolean
  /** 状态描述 */
  detail: string
}

/**
 * OAuth Device Flow 成功事件，对应后端 runtime.EventsEmit("oauth:success", ...)。
 */
export interface OAuthSuccessEvent {
  /** 设备码 */
  deviceCode: string
  /** 厂商类型 */
  providerType: string
  /** 新创建的 Provider ID */
  providerID: string
  /** Provider 显示名称 */
  providerName: string
}

/**
 * Ollama 环境检测结果，对应后端 OllamaDetectResult。
 */
export interface OllamaDetectResult {
  /** ollama 命令是否存在于 PATH */
  installed: boolean
  /** 11434 端口是否响应 */
  running: boolean
  /** smollm2:135m 模型是否已下载 */
  has_smollm2: boolean
  /** 未安装时返回的安装引导文本 */
  install_guide?: string
  /** 正在后台启动服务 */
  server_starting?: boolean
  /** 模型下载进度文本 */
  pull_progress?: string
}

/**
 * 单种认证方式的检测结果，对应后端 AuthMethodDetectStatus。
 */
export interface AuthMethodDetectStatus {
  /** 认证方式：cli_token | oauth_device | api_key | local */
  method: string
  /** 该方式是否可用 */
  available: boolean
  /** 是否已连接/认证成功 */
  connected: boolean
  /** Tier 优先级 1-4 */
  tier: number
  /** 检测到的厂商类型 */
  provider_type?: string
  /** 状态描述文本 */
  detail?: string
  /** 不可用原因 */
  error?: string
}

/**
 * 认证方式统一检测结果，对应后端 AuthDetectResult。
 */
export interface AuthDetectResult {
  /** 各认证方式检测状态列表 */
  results: AuthMethodDetectStatus[]
  /** 推荐的方法 */
  recommended: string
  /** 是否全部不可用 */
  all_unavailable: boolean
}

/**
 * 服务商下的单个模型配置。
 */
export interface ProviderModel {
  /** 模型ID */
  id: string
  /** 显示名称 */
  name: string
  /** 是否启用 */
  enabled: boolean
}

/**
 * 用户已添加的 Provider 配置。
 * 由 ProviderTemplate + 用户输入（认证配置、自定义参数）构成。
 */
export interface ProviderConfig {
  /** 实例唯一ID（模板ID + 时间戳后缀，或自定义UUID） */
  id: string
  /** 来源模板ID */
  templateId: string
  /** 显示名称 */
  name: string
  /** API 基础地址 */
  apiHost: string
  /** API 密钥（api_key 方式下使用；内存中短暂存在，持久化由系统密钥环处理） */
  apiKey: string
  /** 当前选用的默认模型ID（向后兼容） */
  modelId: string
  /** 该服务商下的模型列表 */
  models?: ProviderModel[]
  /** 温度参数，范围 0-2 */
  temperature: number
  /** 请求超时毫秒数 */
  timeoutMs: number
  /** 最大重试次数 */
  maxRetries: number
  /** 分组名称 */
  group: string
  /** 是否启用 */
  enabled: boolean
  /** 排序权重 */
  sortOrder: number
  /** 创建时间戳 */
  createdAt: number
  /** 更新时间戳 */
  updatedAt: number
  /** 认证方式，默认 api_key */
  authMethod: AuthMethod
  /** 认证方式特有参数 */
  authParams: AuthParams
  /** 标记 API Key 是否需补全（导入时留空则标记） */
  needsApiKey?: boolean
}
